# <font color='purple'>**PART-A : Apply Apriori algorithm on the Online retail dataset.**

**Import neccessary libraries and upload the online_retail_II file**

In [ ]:
#uploading the csv file in google colab
import io
import pandas as pd
from google.colab import files
uploaded = files.upload()

Saving B_34_Nidhi Misra - online_retail_II.xlsx to B_34_Nidhi Misra - online_retail_II.xlsx


In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

**Read the file and drop all the nan values**

In [ ]:
df=pd.read_excel("B_34_Nidhi Misra - online_retail_II.xlsx")
df

Invoice StockCode                          Description  Quantity  \
0       536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1       536365     71053                  WHITE METAL LANTERN         6   
2       536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3       536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4       536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...        ...       ...                                  ...       ...   
541905  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
541909  581587      POST                              POSTAGE         1   

               InvoiceDate  Price  Customer ID         Country  
0      2010-12-01 08:26:00   2.55      17850.0  United Kingdom  
1      2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
2      2010-12-01 08:26:00   2.75      17850.0  United Kingdom  
3      2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
4      2010-12-01 08:26:00   3.39      17850.0  United Kingdom  
...                    ...    ...          ...             ...  
541905 2011-12-09 12:50:00   2.10      12680.0          France  
541906 2011-12-09 12:50:00   4.15      12680.0          France  
541907 2011-12-09 12:50:00   4.15      12680.0          France  
541908 2011-12-09 12:50:00   4.95      12680.0          France  
541909 2011-12-09 12:50:00  18.00      12680.0          France  

[541910 rows x 8 columns]

In [ ]:
df=df.dropna()

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406830 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      406830 non-null  object        
 1   StockCode    406830 non-null  object        
 2   Description  406830 non-null  object        
 3   Quantity     406830 non-null  int64         
 4   InvoiceDate  406830 non-null  datetime64[ns]
 5   Price        406830 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      406830 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.9+ MB


In [ ]:
df_plus=df[df['Quantity']>=1]
df_plus.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 397925 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      397925 non-null  object        
 1   StockCode    397925 non-null  object        
 2   Description  397925 non-null  object        
 3   Quantity     397925 non-null  int64         
 4   InvoiceDate  397925 non-null  datetime64[ns]
 5   Price        397925 non-null  float64       
 6   Customer ID  397925 non-null  float64       
 7   Country      397925 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.3+ MB




---


### **a.** **Create basket data using the transactions from Germany, encode it and consider only the transactions with more than 1 item only**.

In [ ]:
basket_germany=(df_plus[df_plus['Country']=="Germany"].groupby(['Invoice','Description'])['Quantity']
             .sum().unstack().reset_index().fillna(0).set_index('Invoice'))
basket_germany

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
Invoice                                                           
536527                                  0.0                 0.0   
536840                                  0.0                 0.0   
536861                                  0.0                 0.0   
536967                                  0.0                 0.0   
536983                                  0.0                 0.0   
...                                     ...                 ...   
581266                                  0.0                 0.0   
581494                                  0.0                 0.0   
581570                                  0.0                 0.0   
581574                                  0.0                 0.0   
581578                                  0.0                 0.0   

Description   I LOVE LONDON MINI BACKPACK   RED SPOT GIFT BAG LARGE  \
Invoice                                                               
536527                                0.0                       0.0   
536840                                0.0                       0.0   
536861                                0.0                       0.0   
536967                                0.0                       0.0   
536983                                0.0                       0.0   
...                                   ...                       ...   
581266                                0.0                       0.0   
581494                                0.0                       0.0   
581570                                0.0                       0.0   
581574                                0.0                       0.0   
581578                                0.0                       0.0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
Invoice                                                                  
536527                                    0.0                      0.0   
536840                                    0.0                      0.0   
536861                                    0.0                      0.0   
536967                                    0.0                      0.0   
536983                                    0.0                      0.0   
...                                       ...                      ...   
581266                                    0.0                      0.0   
581494                                    0.0                      0.0   
581570                                    0.0                      0.0   
581574                                    0.0                      0.0   
581578                                    0.0                      0.0   

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
Invoice                                                           
536527                          0.0                         0.0   
536840                          0.0                         0.0   
536861                          0.0                         0.0   
536967                          0.0                         0.0   
536983                          0.0                         0.0   
...                             ...                         ...   
581266                          0.0                         0.0   
581494                          0.0                         0.0   
581570                          0.0                         0.0   
581574                          0.0                         0.0   
581578                          0.0                         0.0   

Description  12 IVORY ROSE PEG PLACE SETTINGS  \
Invoice                                         
536527                                    0.0   
536840                                    0.0   
536861                                    0.0   
536967                                    0.0   
536983                                    0.0   
...                       

In [ ]:
def encode_units(x):
  if x<=0:
    return 0
  if x>=1:
    return 1
basket_encode_germany=basket_germany.applymap(encode_units)
basket_encode_germany

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
Invoice                                                           
536527                                    0                   0   
536840                                    0                   0   
536861                                    0                   0   
536967                                    0                   0   
536983                                    0                   0   
...                                     ...                 ...   
581266                                    0                   0   
581494                                    0                   0   
581570                                    0                   0   
581574                                    0                   0   
581578                                    0                   0   

Description   I LOVE LONDON MINI BACKPACK   RED SPOT GIFT BAG LARGE  \
Invoice                                                               
536527                                  0                         0   
536840                                  0                         0   
536861                                  0                         0   
536967                                  0                         0   
536983                                  0                         0   
...                                   ...                       ...   
581266                                  0                         0   
581494                                  0                         0   
581570                                  0                         0   
581574                                  0                         0   
581578                                  0                         0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
Invoice                                                                  
536527                                      0                        0   
536840                                      0                        0   
536861                                      0                        0   
536967                                      0                        0   
536983                                      0                        0   
...                                       ...                      ...   
581266                                      0                        0   
581494                                      0                        0   
581570                                      0                        0   
581574                                      0                        0   
581578                                      0                        0   

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
Invoice                                                           
536527                            0                           0   
536840                            0                           0   
536861                            0                           0   
536967                            0                           0   
536983                            0                           0   
...                             ...                         ...   
581266                            0                           0   
581494                            0                           0   
581570                            0                           0   
581574                            0                           0   
581578                            0                           0   

Description  12 IVORY ROSE PEG PLACE SETTINGS  \
Invoice                                         
536527                                      0   
536840                                      0   
536861                                      0   
536967                                      0   
536983                                      0   
...                       

In [ ]:
basket_filter_germany=basket_encode_germany[(basket_encode_germany>0).sum(axis=1)>=2]
basket_filter_germany

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
Invoice                                                           
536527                                    0                   0   
536840                                    0                   0   
536861                                    0                   0   
536967                                    0                   0   
536983                                    0                   0   
...                                     ...                 ...   
581266                                    0                   0   
581494                                    0                   0   
581570                                    0                   0   
581574                                    0                   0   
581578                                    0                   0   

Description   I LOVE LONDON MINI BACKPACK   RED SPOT GIFT BAG LARGE  \
Invoice                                                               
536527                                  0                         0   
536840                                  0                         0   
536861                                  0                         0   
536967                                  0                         0   
536983                                  0                         0   
...                                   ...                       ...   
581266                                  0                         0   
581494                                  0                         0   
581570                                  0                         0   
581574                                  0                         0   
581578                                  0                         0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
Invoice                                                                  
536527                                      0                        0   
536840                                      0                        0   
536861                                      0                        0   
536967                                      0                        0   
536983                                      0                        0   
...                                       ...                      ...   
581266                                      0                        0   
581494                                      0                        0   
581570                                      0                        0   
581574                                      0                        0   
581578                                      0                        0   

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
Invoice                                                           
536527                            0                           0   
536840                            0                           0   
536861                            0                           0   
536967                            0                           0   
536983                            0                           0   
...                             ...                         ...   
581266                            0                           0   
581494                            0                           0   
581570                            0                           0   
581574                            0                           0   
581578                            0                           0   

Description  12 IVORY ROSE PEG PLACE SETTINGS  \
Invoice                                         
536527                                      0   
536840                                      0   
536861                                      0   
536967                                      0   
536983                                      0   
...                       



---


### **b. Apply the apriori algorithm and find associations between the frequently bought items. Conclude which items are bought together.**

In [ ]:
frequent_itemsets_germany=apriori(basket_filter_germany,min_support=0.03,use_colnames=True).sort_values('support',ascending=False).reset_index(drop=True)
frequent_itemsets_germany['length']=frequent_itemsets_germany['itemsets'].apply(lambda x:len(x))
frequent_itemsets_germany

support                                           itemsets  length
0    0.845266                                          (POSTAGE)       1
1    0.258661              (ROUND SNACK BOXES SET OF4 WOODLAND )       1
2    0.237875     (ROUND SNACK BOXES SET OF4 WOODLAND , POSTAGE)       2
3    0.166282               (ROUND SNACK BOXES SET OF 4 FRUITS )       1
4    0.159353      (ROUND SNACK BOXES SET OF 4 FRUITS , POSTAGE)       2
..        ...                                                ...     ...
464  0.030023                                 (STARS GIFT TAPE )       1
465  0.030023                       (STORAGE TIN VINTAGE DOILY )       1
466  0.030023                              (TOADSTOOL MONEY BOX)       1
467  0.030023  (ROUND SNACK BOXES SET OF4 WOODLAND , RETROSPO...       2
468  0.030023                 (JUMBO BAG PINK POLKADOT, POSTAGE)       2

[469 rows x 3 columns]

In [ ]:
association_rules(frequent_itemsets_germany,metric='confidence',min_threshold=0.02).sort_values('lift',ascending=False).reset_index(drop=True)

antecedents                         consequents  \
0              (SPACEBOY CHILDRENS BOWL)            (SPACEBOY CHILDRENS CUP)   
1               (SPACEBOY CHILDRENS CUP)           (SPACEBOY CHILDRENS BOWL)   
2     (POSTAGE, SPACEBOY CHILDRENS BOWL)            (SPACEBOY CHILDRENS CUP)   
3               (SPACEBOY CHILDRENS CUP)  (POSTAGE, SPACEBOY CHILDRENS BOWL)   
4              (SPACEBOY CHILDRENS BOWL)   (POSTAGE, SPACEBOY CHILDRENS CUP)   
...                                  ...                                 ...   
1027                           (POSTAGE)       (SET/6 RED SPOTTY PAPER CUPS)   
1028                (RED KITCHEN SCALES)                           (POSTAGE)   
1029                           (POSTAGE)                (RED KITCHEN SCALES)   
1030               (MINT KITCHEN SCALES)                           (POSTAGE)   
1031                           (POSTAGE)               (MINT KITCHEN SCALES)   

      antecedent support  consequent support   support  confidence       lift  \
0               0.043880            0.046189  0.039261    0.894737  19.371053   
1               0.046189            0.043880  0.039261    0.850000  19.371053   
2               0.041570            0.046189  0.036952    0.888889  19.244444   
3               0.046189            0.041570  0.036952    0.800000  19.244444   
4               0.043880            0.043880  0.036952    0.842105  19.191136   
...                  ...                 ...       ...         ...        ...   
1027            0.845266            0.053118  0.032333    0.038251   0.720124   
1028            0.057737            0.845266  0.034642    0.600000   0.709836   
1029            0.845266            0.057737  0.034642    0.040984   0.709836   
1030            0.055427            0.845266  0.032333    0.583333   0.690118   
1031            0.845266            0.055427  0.032333    0.038251   0.690118   

      leverage  conviction  
0     0.037234    9.061201  
1     0.037234    6.374134  
2     0.035031    8.584296  
3     0.035031    4.792148  
4     0.035026    6.055427  
...        ...         ...  
1027 -0.012566    0.984542  
1028 -0.014161    0.386836  
1029 -0.014161    0.982531  
1030 -0.014518    0.371363  
1031 -0.014518    0.982141  

[1032 rows x 9 columns]

**Analysis:**
  From the above data, we can conclude that if a customer buys **SPACEBOY CHILDRENS BOWL** then there is high probability that he/she will also buy **SPACEBOY CHILDRENS CUP**, since the confidence and lift values are **0.894737** and **19.371053** repectively



---


### **c.Create basket data using the transactions from France and perform the above steps**

In [ ]:
basket_france=(df_plus[df_plus['Country']=="France"].groupby(['Invoice','Description'])['Quantity']
             .sum().unstack().reset_index().fillna(0).set_index('Invoice'))
basket_france

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
Invoice                                                           
536370                                  0.0                 0.0   
536852                                  0.0                 0.0   
536974                                  0.0                 0.0   
537065                                  0.0                 0.0   
537463                                  0.0                 0.0   
...                                     ...                 ...   
580986                                  0.0                 0.0   
581001                                  0.0                 0.0   
581171                                  0.0                 0.0   
581279                                  0.0                 0.0   
581587                                  0.0                 0.0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
Invoice                                                               
536370                                0.0                       0.0   
536852                                0.0                       0.0   
536974                                0.0                       0.0   
537065                                0.0                       0.0   
537463                                0.0                       0.0   
...                                   ...                       ...   
580986                                0.0                       0.0   
581001                                0.0                       0.0   
581171                                0.0                       0.0   
581279                                0.0                       0.0   
581587                                0.0                       0.0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
Invoice                                                                  
536370                                   24.0                      0.0   
536852                                    0.0                      0.0   
536974                                    0.0                      0.0   
537065                                    0.0                      0.0   
537463                                    0.0                      0.0   
...                                       ...                      ...   
580986                                    0.0                      0.0   
581001                                    0.0                      0.0   
581171                                    0.0                      0.0   
581279                                    0.0                      0.0   
581587                                    0.0                      0.0   

Description   TRELLIS COAT RACK  10 COLOUR SPACEBOY PEN  \
Invoice                                                   
536370                      0.0                     0.0   
536852                      0.0                     0.0   
536974                      0.0                     0.0   
537065                      0.0                     0.0   
537463                      0.0                     0.0   
...                         ...                     ...   
580986                      0.0                     0.0   
581001                      0.0                     0.0   
581171                      0.0                     0.0   
581279                      0.0                     0.0   
581587                      0.0                     0.0   

Description  12 COLOURED PARTY BALLOONS  12 EGG HOUSE PAINTED WOOD  ...  \
Invoice                                                             ...   
536370                              0.0                        0.0  ...   
536852                              0.0                        0.0  ...   
536974                              0.0                        0.0  ...   
537065                              0.0                        0.0  ...   
537463                 

In [ ]:
def encode_units(x):
  if x<=0:
    return 0
  if x>=1:
    return 1
basket_encode_france=basket_france.applymap(encode_units)
basket_encode_france

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
Invoice                                                           
536370                                    0                   0   
536852                                    0                   0   
536974                                    0                   0   
537065                                    0                   0   
537463                                    0                   0   
...                                     ...                 ...   
580986                                    0                   0   
581001                                    0                   0   
581171                                    0                   0   
581279                                    0                   0   
581587                                    0                   0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
Invoice                                                               
536370                                  0                         0   
536852                                  0                         0   
536974                                  0                         0   
537065                                  0                         0   
537463                                  0                         0   
...                                   ...                       ...   
580986                                  0                         0   
581001                                  0                         0   
581171                                  0                         0   
581279                                  0                         0   
581587                                  0                         0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
Invoice                                                                  
536370                                      1                        0   
536852                                      0                        0   
536974                                      0                        0   
537065                                      0                        0   
537463                                      0                        0   
...                                       ...                      ...   
580986                                      0                        0   
581001                                      0                        0   
581171                                      0                        0   
581279                                      0                        0   
581587                                      0                        0   

Description   TRELLIS COAT RACK  10 COLOUR SPACEBOY PEN  \
Invoice                                                   
536370                        0                       0   
536852                        0                       0   
536974                        0                       0   
537065                        0                       0   
537463                        0                       0   
...                         ...                     ...   
580986                        0                       0   
581001                        0                       0   
581171                        0                       0   
581279                        0                       0   
581587                        0                       0   

Description  12 COLOURED PARTY BALLOONS  12 EGG HOUSE PAINTED WOOD  ...  \
Invoice                                                             ...   
536370                                0                          0  ...   
536852                                0                          0  ...   
536974                                0                          0  ...   
537065                                0                          0  ...   
537463                 

In [ ]:
basket_filter_france=basket_encode_france[(basket_encode_france>0).sum(axis=1)>=2]
basket_filter_france

Description   50'S CHRISTMAS GIFT BAG LARGE   DOLLY GIRL BEAKER  \
Invoice                                                           
536370                                    0                   0   
536852                                    0                   0   
536974                                    0                   0   
537065                                    0                   0   
537463                                    0                   0   
...                                     ...                 ...   
580986                                    0                   0   
581001                                    0                   0   
581171                                    0                   0   
581279                                    0                   0   
581587                                    0                   0   

Description   I LOVE LONDON MINI BACKPACK   NINE DRAWER OFFICE TIDY  \
Invoice                                                               
536370                                  0                         0   
536852                                  0                         0   
536974                                  0                         0   
537065                                  0                         0   
537463                                  0                         0   
...                                   ...                       ...   
580986                                  0                         0   
581001                                  0                         0   
581171                                  0                         0   
581279                                  0                         0   
581587                                  0                         0   

Description   SET 2 TEA TOWELS I LOVE LONDON    SPACEBOY BABY GIFT SET  \
Invoice                                                                  
536370                                      1                        0   
536852                                      0                        0   
536974                                      0                        0   
537065                                      0                        0   
537463                                      0                        0   
...                                       ...                      ...   
580986                                      0                        0   
581001                                      0                        0   
581171                                      0                        0   
581279                                      0                        0   
581587                                      0                        0   

Description   TRELLIS COAT RACK  10 COLOUR SPACEBOY PEN  \
Invoice                                                   
536370                        0                       0   
536852                        0                       0   
536974                        0                       0   
537065                        0                       0   
537463                        0                       0   
...                         ...                     ...   
580986                        0                       0   
581001                        0                       0   
581171                        0                       0   
581279                        0                       0   
581587                        0                       0   

Description  12 COLOURED PARTY BALLOONS  12 EGG HOUSE PAINTED WOOD  ...  \
Invoice                                                             ...   
536370                                0                          0  ...   
536852                                0                          0  ...   
536974                                0                          0  ...   
537065                                0                          0  ...   
537463                 

In [ ]:
frequent_itemsets_france=apriori(basket_filter_france,min_support=0.03,use_colnames=True).sort_values('support',ascending=False).reset_index(drop=True)
frequent_itemsets_france['length']=frequent_itemsets_france['itemsets'].apply(lambda x:len(x))
frequent_itemsets_france

support                                           itemsets  length
0    0.793566                                          (POSTAGE)       1
1    0.195710                               (RABBIT NIGHT LIGHT)       1
2    0.187668                    (RED TOADSTOOL LED NIGHT LIGHT)       1
3    0.179625                 (PLASTERS IN TIN WOODLAND ANIMALS)       1
4    0.176944                   (PLASTERS IN TIN CIRCUS PARADE )       1
..        ...                                                ...     ...
782  0.032172  (SPACEBOY LUNCH BOX , LUNCH BAG RED RETROSPOT,...       3
783  0.032172  (POSTAGE, LUNCH BAG WOODLAND, DOLLY GIRL LUNCH...       3
784  0.032172  (ROUND SNACK BOXES SET OF4 WOODLAND , POSTAGE,...       3
785  0.032172  (POSTAGE, WOODLAND MINI BACKPACK, DOLLY GIRL M...       3
786  0.032172  (SET/6 RED SPOTTY PAPER CUPS, PACK OF 6 SKULL ...       5

[787 rows x 3 columns]

In [ ]:
association_rules(frequent_itemsets_france,metric='lift',min_threshold=1).sort_values('lift',ascending=False).reset_index(drop=True)

antecedents  \
0     (PACK OF 20 SKULL PAPER NAPKINS, SET/6 RED SPO...   
1     (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...   
2     (SET/6 RED SPOTTY PAPER CUPS, PACK OF 6 SKULL ...   
3     (SET/6 RED SPOTTY PAPER CUPS, PACK OF 6 SKULL ...   
4     (POSTAGE, PACK OF 20 SKULL PAPER NAPKINS, SET/...   
...                                                 ...   
2789  (DOLLY GIRL CHILDRENS CUP, DOLLY GIRL CHILDREN...   
2790  (LUNCH BAG APPLE DESIGN, LUNCH BAG SPACEBOY DE...   
2791                                          (POSTAGE)   
2792                              (SPACEBOY LUNCH BOX )   
2793                                          (POSTAGE)   

                                            consequents  antecedent support  \
0     (SET/20 RED RETROSPOT PAPER NAPKINS , SET/6 RE...            0.037534   
1     (PACK OF 20 SKULL PAPER NAPKINS, SET/6 RED SPO...            0.034853   
2     (PACK OF 20 SKULL PAPER NAPKINS, SET/6 RED SPO...            0.042895   
3     (PACK OF 6 SKULL PAPER CUPS, PACK OF 20 SKULL ...            0.042895   
4     (SET/6 RED SPOTTY PAPER CUPS, PACK OF 6 SKULL ...            0.034853   
...                                                 ...                 ...   
2789                                          (POSTAGE)            0.040214   
2790                                          (POSTAGE)            0.067024   
2791  (LUNCH BAG APPLE DESIGN, LUNCH BAG SPACEBOY DE...            0.793566   
2792                                          (POSTAGE)            0.131367   
2793                              (SPACEBOY LUNCH BOX )            0.793566   

      consequent support   support  confidence       lift  leverage  \
0               0.034853  0.032172    0.857143  24.593407  0.030863   
1               0.037534  0.032172    0.923077  24.593407  0.030863   
2               0.037534  0.037534    0.875000  23.312500  0.035923   
3               0.032172  0.032172    0.750000  23.312500  0.030792   
4               0.042895  0.034853    1.000000  23.312500  0.033358   
...                  ...       ...         ...        ...       ...   
2789            0.793566  0.032172    0.800000   1.008108  0.000259   
2790            0.793566  0.053619    0.800000   1.008108  0.000431   
2791            0.067024  0.053619    0.067568   1.008108  0.000431   
2792            0.793566  0.104558    0.795918   1.002965  0.000309   
2793            0.131367  0.104558    0.131757   1.002965  0.000309   

      conviction  
0       6.756032  
1      12.512064  
2       7.699732  
3       3.871314  
4            inf  
...          ...  
2789    1.032172  
2790    1.032172  
2791    1.000583  
2792    1.011528  
2793    1.000449  

[2794 rows x 9 columns]



---

### **d. Use appropriate visualization and show the comparison between different countries to conclude. For example, compare the most frequently purchased item in different countries.**